Implemented DCT decompression

wolfhechel opened the pull request:
Added a functional DCT decompressor instead of using dct.dll, have not been able to test Unpack_Bloc properly.

There's also a commit which implements some fixes needed for compiling on Linux.

wolfhechel committed:
Own DCT implementation.

wolfhechel commented:
@Alys-Elica this isn't quite ready for merge yet, but it works and I wanted to keep you in the loop and get some early feedback.

I'm much more of a C developer than C++ so if I'm doing anything weird could you perhaps leave a comment there?

Also, I haven't finished writing up how the DCT works, I'll add some documentation and more comments before merging.

One thing you'll notice if you compare this implementation with the previous one is that the color saturations are way higher. My suspicion is that the floating to long conversion is different from that of msvcrt, the change happens in the YCbCr -> RGBA conversion.

Also I'm thinking that perhaps its a good idea to retain the dll based implementation that was already in place and male it conditional, what are your thoughts on that?

Alys-Elica commented on Tools/LibFVR/internal/dct.cpp:
```diff
@@ -1,107 +1,598 @@
+           we'll keep an extra width size for the overflowing bits which
+           will pass on to the next iteration. */
+        
+        int *r = (int *)malloc(sizeof(int[width * 9])),
+            *g = (int *)malloc(sizeof(int[width * 9])),
+            *b = (int *)malloc(sizeof(int[width * 9]));
```
Doesn't compile on MSVC (error C2540)

Alys-Elica commented on Tools/LibFVR/internal/dct.cpp:
```diff
@@ -1,107 +1,598 @@
-        DCT_init = (DCT_cdtor)GetProcAddress(libDct, "??0DCT@@QAE@XZ");
-        DCT_deinit = (DCT_cdtor)GetProcAddress(libDct, "??1DCT@@QAE@XZ");
-        if (!DCT_init || !DCT_deinit)
+        buffer[513 * 3] = 0x7fffffff;
```
Index is out of bounds

Alys-Elica reviewed (changes requested):
When the above are fixed, the code still crashes (out of bound access)

Alys-Elica commented:
> @Alys-Elica this isn't quite ready for merge yet, but it works and I wanted to keep you in the loop and get some early feedback.
> 
> I'm much more of a C developer than C++ so if I'm doing anything weird could you perhaps leave a comment there?
> 
> Also, I haven't finished writing up how the DCT works, I'll add some documentation and more comments before merging.
> 
> One thing you'll notice if you compare this implementation with the previous one is that the color saturations are way higher. My suspicion is that the floating to long conversion is different from that of msvcrt, the change happens in the YCbCr -> RGBA conversion.
> 
> Also I'm thinking that perhaps its a good idea to retain the dll based implementation that was already in place and male it conditional, what are your thoughts on that?

Saw your comment a bit late, sorry, I was writing my change request at the same time.

For the C/C++ "weirdiness", the only thing I see for now is the use of malloc. For raw data storage, I like to use std::vector<uint8_t> witch also takes care of the memory management.
There is also already a std::clamp function that exists since C++17.
Also, beware of the unfreed `new`ed objects, they need to be `delete`d when not used anymore. For the objects you use (HuffmanTable and InverseBitStream), I think you could just work with classic non-pointer objects.
If you need infos, you can check the [CPP Reference](cppreference.com) website, it is quite a good doc.

My take on keeping the DLL based DCT is that it could be useful for testing/debugging. For me, the problem using this method is that it locks us in a 32bit Windows only 32bit version. But you are right, keeping the two for now can be useful.

Finally, as I noted previously, I was unable to make it work on my setup, so I didn't fully review it.

wolfhechel committed:
Conditional compilation for DLL dependent parts

We can safely compile this program under Linux but without support for
LoadLibrary. Meaning we'll lose the vr-viewer support until a differnt
VR viewer has been implemented.

wolfhechel committed:
Fixed a segfault where memset would overflow.

When clearing the first width * 8 of r, g and b respectively I was
including the int size when calculating the offset. This is not
necessary as the pointer is incremented by its typesize.

wolfhechel committed:
Add additional byte to work buffer when unpacking huffman table.

wolfhechel committed:
Fixed up some issues with the cross-platform compilation when compiling
under mingw for windows.

wolfhechel committed:
Accidentally used luminance quants in DCT, fixes oversaturation.

wolfhechel committed:
Merge branch 'main' into dct

* main:
  [Tools] Reworked FnxVr a bit
  [Tools] Replaced the Fnx_vr renderer by a custom one
  [Tools] Added 90 degrees rotation in VR viewer
  [Tools] Fixed FnxVr angle clamping
  [Doc] Added link to LST doc in main readme
  [Tools] Added TST file loading
  Documented the TST file format and started describing the LST script syntax.

wolfhechel committed:
Need to include cstdint and cmath for GCC.

std::fabs and uint16_t are declared in these headers, needed for
compilation to succeed.

wolfhechel committed:
Remove windows.h dependency check

Since the VR renderer no longer loads a DLL, this check is pointless.
